fluss-flink/fluss-flink-common/src/main/java/com/alibaba/fluss/flink/sink/writer/FlinkSinkWriter.java

```diff
@@ -120,29 +159,40 @@ protected void initMetrics() {
     }
 
     @Override
-    public void write(RowData value, Context context) throws IOException, InterruptedException {
+    public void write(InputT inputValue, Context context) throws IOException, InterruptedException {
         checkAsyncException();
-        if (ignoreDelete
-                && (value.getRowKind() == RowKind.UPDATE_BEFORE
-                        || value.getRowKind() == RowKind.DELETE)) {
-            return;
-        }
 
-        InternalRow internalRow = sinkRow.replace(value);
-        CompletableFuture<?> writeFuture = writeRow(value.getRowKind(), internalRow);
-        writeFuture.whenComplete(
-                (ignored, throwable) -> {
-                    if (throwable != null) {
-                        if (this.asyncWriterException == null) {
-                            this.asyncWriterException = throwable;
-                        }
+        try {
+            RowWithOp rowWithOp = serializationSchema.serialize(inputValue);
+
+            InternalRow internalRow = rowWithOp.getInternalRow();
+            RowKind rowKind = rowWithOp.getRowKind();
 
-                        // Checking for exceptions from previous writes
-                        mailboxExecutor.execute(this::checkAsyncException, "Update error metric");
-                    }
-                });
+            if (ignoreDelete && (rowKind == RowKind.UPDATE_BEFORE || rowKind == RowKind.DELETE)) {
```

Comment:
@wuchong although i made some progress, seems like UPSERTS are not only handled on the server.. here we need to check if the RowKind is an `UPDATE_BEFORE` so we can handle the `ignoreDelete`